docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,239 @@
+mentioned some cases like web-based explorer UIs and possibly other tools.
+
+As far as I can tell the primary consumers of the multi-headed "RPC service"
+are the light client and the `tendermint` command-line client. There is
+probably some local use via curl, but I expect that is mostly ad hoc. Ethan
+reports that nodes are often configured with the ports to the RPC service
+blocked, which is good for security but complicates use by the light client.
+
+TODO: Figure out what is going on with the weird pseudo-inheritance service
+interface.  The BaseService type makes sense, but its embeds are more
+complicated.
+
+### Context: ABCI Issues
+
+In the original design of ABCI, the presumption was that all access to the
+application would be mediated by the consensus node. This makes sense, since
+outside access could change application state and corrupt the consensus
+process. Of course, even without outside access an application could behave
+nondeterministically, but allowing other programs to send it requests was seen
+as inviting trouble.
+
+Later, however, it was noted that most of the time, tools specific to a
+particular application don't really want to talk to the consensus module
+directly. The application is the "owner" of the state machine the consensus
+engine is replicating, so tools that want to know things about application
+state should talk to the application. Otherwise, such tools would have to bake
+in knowledge about Tendermint (e.g., its interfaces and data structures) that
+they might otherwise not need or care about.
+
+However, this raises another issue: The consensus node is the ABCI client, so
+it is inconvenient for the application to "push" work into the consensus
+module. In theory you could work around this by having the consensus module
+"poll" the application for work that needs done, but that has unsatisfactory
+implications for performance and robustness, as well as being harder to
+understand.
+
+There has apparently been some discussion about trying to make a more
+bidirectional communication between the consensus node and the application, but
+this issue seems to still be under discussion. TODO: Impact of ABCI++ for this
+question?
+
+### Context: RPC Issues
+
+The RPC system serves several masters, and has a complex surface area. I
+believe there are some improvements that can be exposed by separating some of
+these concerns.
+
+For light clients, some work is already underway to move toward using P2P
+message passing rather than RPC to synchronize light client state with the rest
+of the network.  Because of the way light client synchronization interacts with
+event logging, this is not as simple as just swapping out the transport: More
+details of the event system will need to be made visible via P2P, and that in
+turn has some implications for fault tolerance (in particular, I think the
+ability of a badly-behaved client to DDoS the network without punishment is a
```

Comment:
I'm assuming the concern here is around a malicious client being able to overwhelm DB queries/ connections to the events DB?

I'm wondering how far you've looked to see the interlinking of the light client with the event system? The light clients need to retrieve and verify blocks at different heights and I'm not yet sure how much that depends on the event system. I see calls for getting a commit at a height  and calls to get a validator set at a height, but these rely on the state and block store.

Reply:
> I'm assuming the concern here is around a malicious client being able to overwhelm DB queries/ connections to the events DB?

I'm not sure, but I believe the specific concern is that exposing the event system on the P2P network makes it reachable by _any_ node in the network, whereas now it's only accessible (I think) locally? That may be wrong. It seems to me this should already be a concern, but perhaps right now it's not a very big target. Relying on the event facility more for synchronization would presumably make it a more attractive attack surface.

My uncertainty about these questions is part of why I still have this marked as a draft. Your questions are all very good, though, so please keep them coming. I will do my best to clarify them. 🙂 

> I'm wondering how far you've looked to see the interlinking of the light client with the event system? 

Not far at all. This claim is mostly based on what @ebuchman described to us, and some of the PRs that have gone by making changes to that system.

Reply:
With @cmwaters's recent changes, the light client does now use P2P for block synchronization, but it appears to still rely on the RPC service (I think via a websocket? 😝) to hook into the event system. 

Without proof, I think we could probably convert the light client away from websocket, even if we don't have a way to completely wean it from the RPC service. That's not the most important problem to solve, but it suggests we can do most of the proposed simplifications without blocking on P2P event access.

Reply:
Bettering light client UX around users, not only using them to verify data is a huge want. Right now, light clients in tendermint aren't really full light clients, but half-baked and only used for verification. Light clients are usually used for so much more, but getting it off the RPC is a must IMO. Putting it on the RPC layer was a mistake in the first place.

Reply:
> but it appears to still rely on the RPC service (I think via a websocket? 😝) to hook into the event system.

There is the light client which just verifies light block and then their is a light node (I'm making up this distinction) in which an RPC server is quasi wrapped around the light client. I think the websocket stuff you are seeing relates to the light node as a server and not the light node as a client. The light node as a client has no relation with the events system (yet). I think @ebuchman suggested pairing the two to benefit IBC relayers who basically need to listen to the network to see when a block is committed and then verify the header (using the light client)

Reply:
> Right now, light clients in tendermint aren't really full light clients, but half-baked and only used for verification. Light clients are usually used for so much more

This is helpful context—I wasn't aware that light clients were a more general thing. Is this something that other ecosystems use for observability/introspection?

> [G]etting it off the RPC is a must IMO. Putting it on the RPC layer was a mistake in the first place.

Agreed. I'd like to move toward a state where the consensus node RPC service is basically a local control and debugging port, rather than a load-bearing part of the consensus network.

Reply:
I updated the language here hoping to capture some of these issues.